### ISSUE TITLE
[runtimes] Fix AnimationState dipping problem

### ISSUE THREAD

**Issue by badlogic** [EsotericSoftware/spine-runtimes#815]
#See http://esotericsoftware.com/forum/Blend-Weirdness-in-Unity-3-5-7263?&hilit=dipping

Includes 3 gists by Nate, 2nd Gist is the closest to a solution. Needs a fix so 3 way mixing does not jump.

Test rig:

```
{
"skeleton": { "hash": "9H74t+0fdSoMflQP/bYFClZz+Xw", "spine": "99.9.99-dev", "width": 0, "height": 0, "fps": 1, "images": "C:/test/" },
"bones": [
	{ "name": "root" },
	{ "name": "bone1", "parent": "root", "length": 200, "x": 50 },
	{ "name": "bone2", "parent": "root", "length": 200, "x": 300 },
	{ "name": "bone3", "parent": "root", "length": 200, "rotation": 0.46, "x": 550 }
],
"animations": {
	"bone1": {
		"bones": {
			"bone1": {
				"rotate": [
					{ "time": 0, "angle": 90, "curve": "stepped" },
					{ "time": 2, "angle": 90 }
				]
			}
		}
	},
	"bone2": {
		"bones": {
			"bone2": {
				"rotate": [
					{ "time": 0, "angle": 90, "curve": "stepped" },
					{ "time": 2, "angle": 90 }
				]
			}
		}
	},
	"bone3": {
		"bones": {
			"bone3": {
				"rotate": [
					{ "time": 0, "angle": 90, "curve": "stepped" },
					{ "time": 2, "angle": 90 }
				]
			}
		}
	},
	"bone1-90": {
		"bones": {
			"bone1": {
				"rotate": [
					{ "time": 0, "angle": -90, "curve": "stepped" },
					{ "time": 2, "angle": -90 }
				]
			}
		}
	},
	"bone1-180": {
		"bones": {
			"bone1": {
				"rotate": [
					{ "time": 0, "angle": 179, "curve": "stepped" },
					{ "time": 2, "angle": 179 }
				]
			}
		}
	},
	"bone1-dupe": {
		"bones": {
			"bone1": {
				"rotate": [
					{ "time": 0, "angle": 90, "curve": "stepped" },
					{ "time": 2, "angle": 90 }
				]
			}
		}
	}
}
}
```

**Comment by badlogic** [comment]
Possibly related http://esotericsoftware.com/forum/IK-Mix-over-Multiple-Animations-in-Quick-Succession-7779?p=36255#p36255

**Comment by badlogic** [comment]
More data in http://esotericsoftware.com/forum/IK-Mix-over-Multiple-Animations-in-Quick-Succession-7779

**Comment by taigacon** [comment]
@badlogic I found that 

> set the mix time between the animations "over" and "over2" to 0

seems not working.

Here is the spine project:
https://drive.google.com/open?id=0B6F3biFZyXZaaVg2bHBZd3MwMFE

In the project there are two animations, 'animation' and 'state'.
Then I used
`setMix("animation","state",0)`
`setAnimation(0,"animation",false)`
`addAnimation(0, "state", true)`
but found that the dip problem is still there.

Here is the video:
https://drive.google.com/open?id=0B6F3biFZyXZaaGRfeHMtTjhrdGs


---------------------------------------------------


Edit:

Sure, by chance I found that the dip problem is fixed only if the two animations' root bones are keyed in the first frame with your solution.


--------------------------------------------------

Edit 2:

Solution in Edit 1 seems not always working for every project...
Here is another project:
https://drive.google.com/open?id=0B6F3biFZyXZaek1SNkpJVFk1X0E

**Comment by pharan** [comment]
Note the following experimental implementations:
Java/libGDX: https://gist.github.com/NathanSweet/3aaeb489c2e8cdca76b31f5ae87f2b8b
C#: https://gist.github.com/pharan/cffdc3b2aee9288c9ca8b4ab94ee97d2

**Comment by pharan** [comment]
The C# version has one change not present in the libGDX version:

replaced
```csharp
float a = setupPose && !timelinesLastItems[i] ? alphaFull : alphaMix;
```

with
```csharp
bool isBasisPose = setupPose && !timelinesLastItems[i] && alphaMix > 0;
float a = isBasisPose ? alphaFull : alphaMix;
```

**Comment by imyourm8** [comment]
@pharan On commenting line 656 all my dipping is gone.

**Comment by pharan** [comment]
@imyourm8 You mean it still had dipping while 656 is uncommented?
I'll have to discuss with @NathanSweet what exactly that line is trying to do.

**Comment by imyourm8** [comment]
@pharan Yes it still appears. When when u play Anim1 -> Anim2 -> Anim1, after Anim2 almost finished i set Anim1 and there is dipping. But commenting out will brake animation if next animation is some cases again. I think when there is some timelines in next animation which not presented in previous.

**Comment by imyourm8** [comment]
Any updates from devs?

**Comment by pharan** [comment]
@imyourm8 Just took a look at it today and made some changes. Try the latest experimental AnimationState: https://gist.githubusercontent.com/pharan/cffdc3b2aee9288c9ca8b4ab94ee97d2/raw/65872ae7e970796bf2b9ddf26483d9baf914b5d7/AnimationState.cs

**Comment by NathanSweet** [comment]
This has been fixed (currently in the `3.6-beta` branch) by by making multiple mixing optional and disabled by default:
13c34e8751aedd48d7914c0e0ff58a322424c18e
You can now choose between multiple mixing with the dipping problem, or no multiple mixing and no dipping problem.
